### ISSUE TITLE
Labels not showing in Roundcube

### ISSUE THREAD

**Issue by mkerstner** [mike-kfed/roundcube-thunderbird_labels#80]
Using RC 1.6.5 with PHP 8.2.x here.

While the plugin works to set the labels from both sides (TB and RC) unfortunately they only show in TB.

I've checked if they show in Classic and Larry -> in both they are only visible once when setting the label in RC.
But they are gone once the folder is refreshed.

There are no JS or PHP erros shown.

Could you provide any hints as to what might cause the issue here? Thanks a bunch for the plugin!

FYI:
To get rid of PHP deprecation warnings I've added the following lines in the main PHP file:

`class thunderbird_labels extends rcube_plugin
{
	public $task = 'mail|settings';
	private $rc;
	private $map;
	private $_custom_flags_allowed = null;
	const LABEL_STYLES = ['thunderbird', 'bullets', 'badges'];

        // ADDED THESE LINED TO GET RID OF PHP 8.2.x WARNINGS	
	private $add_tb_flags;
	private $name;
	private $message_tb_labels;`

**Comment by mkerstner** [comment]
Update:

With further testing I can confirm that labels are showing correctly for all folders except INBOX.

Once I move an already flagged email from INBOX to any sub folder RC will show the TB flags correctly.

But they don't show at all in INBOX.

Anything that I can test here to provide further insights / solution paths?

Thanks

**Comment by mkerstner** [comment]
Are you using dovecot by any chance? If so, could you please share your configuration? I would check it from this perspective and then continue debugging. Thanks

**Comment by mike-kfed** [comment]
I am using dovecot 2.3 with maildir storage, the config is the default that comes with Debian 12, the only thing I adapted are the user lookups, they are coming from a sql database. But the latter has no impact on how the inbox is handled, so sharing my config is not needed, just inspect the .deb file.

**Comment by c-schwamborn** [comment]
I think the issue is in `function set_flags`, on my system (bookworm with stock plugins) around line 318 (328 in the current git code). there is a definition `$imap = $this->rc->imap;`. The object `$this->rc->imap` seems to be renamed to `$this->rc->storage`. Changing this one line fixed the issue for me.
Otherwise I got an error like this, trying to set a message Flag:
`[05-Apr-2024 15:09:03 UTC] PHP Warning:  Undefined property: rcmail::$imap in /usr/share/roundcube/plugins/thunderbird_labels/thunderbird_labels.php on line 318`
`[05-Apr-2024 15:09:03 UTC] PHP Fatal error:  Uncaught TypeError: array_merge(): Argument #1 must be of type array, null given in /usr/share/roundcube/plugins/thunderbird_labels/thunderbird_labels.php:327`

**Comment by mike-kfed** [comment]
> I think the issue is in `function set_flags`, on my system (bookworm with stock plugins) around line 318 (328 in the current git code). there is a definition `$imap = $this->rc->imap;`. The object `$this->rc->imap` seems to be renamed to `$this->rc->storage`.

You are clearly not using the latest thunderbird lables release 1.6.1 https://github.com/mike-kfed/roundcube-thunderbird_labels/blob/77b59ebb606c9185a2e2b8ee339ed13b2c36e1fb/thunderbird_labels.php#L325 this line was fixed in january 2023 to use storage instead of imap. I am not sure how your comment is helping with the issue at hand?

**Comment by c-schwamborn** [comment]
Yeah, sorry for the post. I was a bit tired when I compared the code
from the debian package with the git code. Later I discovered my
mistake. Sadly the maintainer know about the patch but, for what ever
reason, leave the current bookworm stable packages in that broken state.
So discard my comment :-)

**Comment by InputOutputZ** [comment]
> > I think the issue is in `function set_flags`, on my system (bookworm with stock plugins) around line 318 (328 in the current git code). there is a definition `$imap = $this->rc->imap;`. The object `$this->rc->imap` seems to be renamed to `$this->rc->storage`.
> 
> You are clearly not using the latest thunderbird labels release 1.6.1
> 
> [roundcube-thunderbird_labels/thunderbird_labels.php](https://github.com/mike-kfed/roundcube-thunderbird_labels/blob/77b59ebb606c9185a2e2b8ee339ed13b2c36e1fb/thunderbird_labels.php#L325)
> 
> Line 325 in [77b59eb](/mike-kfed/roundcube-thunderbird_labels/commit/77b59ebb606c9185a2e2b8ee339ed13b2c36e1fb)
> 
>  $imap = $this->rc->storage; 
> this line was fixed in january 2023 to use storage instead of imap. I am not sure how your comment is helping with the issue at hand?

Actually there seem to be a bug here when you enable $config['messages_cache'] = true, this plugin can only display the 5 custom labels, and fails to display any flag added through from somewhere e.g. through sieve.

Here are a screenshots showcasing difference to an email has flag already set to

When you enable messages_cache
<img width="1075" alt="Image" src="https://github.com/user-attachments/assets/b08e5635-8cbc-4513-bc45-45ebfa9af477" />

When you disable messages_cache
<img width="1074" alt="Image" src="https://github.com/user-attachments/assets/041151fa-2fa5-4329-927b-b317fad12719" />

I would love to get this fixed but not sure where to start debugging? Any suggestion would be very much appreciated.

With thanks.

**Comment by mkerstner** [comment]
I can confirm that disabling the messages cache fixes my issue as well. (Not ideal, but at least the labels are showing now).

+1 on supporting to debug -> any hints?

**Comment by mike-kfed** [comment]
I don't know what `messages_cache` does, but I know that supporting custom labels is in a half working state and wont be finished until roundcube itself gets it's act together not mangling label names, see discussion #43  

So you are seeing something "working" that right now is out of scope of this plugin and works by random chance. Feel free to patch to roundcube and this plugin to make it work though :)

**Comment by InputOutputZ** [comment]
> I don't know what `messages_cache` does, but I know that supporting custom labels is in a half working state and wont be finished until roundcube itself gets it's act together not mangling label names, see discussion [#43](https://github.com/mike-kfed/roundcube-thunderbird_labels/issues/43)
> 
> So you are seeing something "working" that right now is out of scope of this plugin and works by random chance. Feel free to patch to roundcube and this plugin to make it work though :)

messages_cache caches messages in db, to load them faster, without having to contact imap. Enable it and open few messages, and open database messages cache table. You will see it with cached messages.

**Comment by mkerstner** [comment]
Ok, so the assumption would be that the caching mechanism does not work as expected in conjunction with custom labels.
Did someone already have look at the db entries / structure to find out what the culprit could be? If not I'll try to free up some time to check it.

**Comment by InputOutputZ** [comment]
> Ok, so the assumption would be that the caching mechanism does not work as expected in conjunction with custom labels. Did someone already have look at the db entries / structure to find out what the culprit could be? If not I'll try to free up some time to check it.

I did, and this is how it looks like.

<img width="1383" alt="Image" src="https://github.com/user-attachments/assets/3e1b5bd8-9a0e-4f93-a6e4-4ee942b4f104" />

The failure reason seems to be because this plugin uses different method than RC to fetch the message which contains flags/custom labels and hence it fails to load flags or it could be same method but cant understand cached content format, I think this plugin needs to check if cached messages is enabled, and use RC method to decompress/decode cached message and fetch the flags/custom labels and I don't think they are mangled, because if they were, how CC flag was fetched in message view, it just seems like the hook this plugin uses to append the flags/custom labels in message view its either before RC assign flags content to a variable or after the variables its being nullified as considered of no use by RC and this plugins needs to perhaps refetch something as it consider it of use at specific point in time and maybe just does what RC did upuntil before destroyed/nullified custom labels/flags content, and here I think there is nothing being mangled, its just how RC works and optimise its code, and the plugins need to workaround this and it is not the other way round, where RC is being redefined for plugins.

I don't I will work on this issue, because I'm the kind of person, will try to understand everything and can't do this now, as it needs a bit of time. If anyone has this time now, it would be very much appreciated if they look at it.
